### ISSUE TITLE
Content Security Policy blocks unsafe usage of eval.

### ISSUE THREAD

**Issue by aglassman** [fidr/phoenix_live_react#25]
I'm having a heck of a time getting a solid Content Security Policy written for a phoenix / LiveView / React app I'm working on.
I have basically everything ironed out, except one last bit concerning phoenix_live_react.

https://github.com/fidr/phoenix_live_react/blob/1ecae366f9a61520f0aa8e903ee45f1660886db4/assets/js/phoenix_live_react.js#L18

This line is using eval, and the CSP will not allow it unless I specify `unsafe-eval` which essentially negates the whole CSP.  Any guidance here?  It seems like it's only being used to turn a string into a react class.  Is there some work around for this to avoid using eval?

![Screenshot 2023-01-13 at 4 23 02 PM](https://user-images.githubusercontent.com/1435196/212430304-4c6b19fb-097c-439d-a46b-775b92ebb05b.png)

**Comment by jeregrine** [comment]
Yea its so they can use the ReactClassName into a proper class, but that is unnecessary, you can simply pas "div" and get the same result here:
https://github.com/fidr/phoenix_live_react/blob/master/assets/js/phoenix_live_react.js#L11

@aglassman you should be able to solve this by removing eval and changing ^ to 

```javascript
  const reactElement = React.createElement("div", props);
```

**Comment by aglassman** [comment]
Thanks @jeregrine. I forked, and will I'll give this a try.

**Comment by aglassman** [comment]
This worked, I created a fork and could create a PR to modify the original library if we think this won't break existing functionality in certain cases.  I also upped the react peer dependency to 17 as that is the version we are using. Not sure if this was necessary or not.  Will create a PR if maintainers are OK with this change.  https://github.com/KGG-Inc/phoenix_live_react

**Comment by fidr** [comment]
It's been a while, but the `eval` was in there to look up the react component in the window scope. e.g. I don't think `"div"` will actually render the correct react element.

@aglassman If you like, can you look at this change and see if that works as well: https://github.com/geolessel/react-phoenix/commit/e46a802e1d31657b741e25fc9608ec85cfee9ef0

I'd happily accept a PR for that!

**Comment by aglassman** [comment]
@fidr Thanks, I'll take a look at that.  It does seem to work when I changed to `"div"`.  How this works is beyond my current understanding of the whole component lifecycle, so not sure if it just happens to work for me and maybe will not work for others.  I'm using React 17, not sure if that makes a difference.

**Comment by jeregrine** [comment]
@aglassman basically change line 18 to 
```js
const componentClass = Array.prototype.reduce.call(e.dataset.reactClass.split('.'), (acc, el) => { return acc[el] }, window);
```

Which should roughly do the same thing and it might make the ReactDeveloperTools work better? Is my guess.

@fidr if you read the docs it accepts a Class or a tag name, its not clear why it accepts both.

**Comment by aglassman** [comment]
Ok, after some testing, I think changing to `"div"` actually only kept things from crashing, and did not actually load the react component correctly.  Testing this with CSP was pretty tricky to test locally as the webpack server requires `'unsafe-inline'`, so I had to spend some time getting things setup right.  In the end, you do actually need the reduce call with split.  I have that working locally, and will create a PR for this change.   Thanks for everyone's help on this.

**Comment by aglassman** [comment]
Created PR #26

**Comment by fidr** [comment]
Fixed in #26
